codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,203 @@
+            State state = emptyFrameQueue.remove();
+            state.enqueued = false;
+            state.write(0, writer);
+        }
+
+        final int size = queue.size();
+        if (size == 0 || maxBytes <= 0) {
+            return totalStreamableBytes > 0;
+        }
+
+        final int chunkSize = max(minAllocationChunk, maxBytes / size);
+
+        // Write until the queue is empty or we've exhausted maxBytes. We need to check queue.isEmpty()
+        // here since the Writer could call updateStreamableBytes, which may alter the queue.
+        do {
+            // Remove the head of the queue.
+            State state = queue.remove();
+            state.enqueued = false;
+
+            // Allocate as much data as we can for this stream.
+            int chunk = min(chunkSize, min(maxBytes, state.streamableBytes));
+            maxBytes -= chunk;
+
+            // Write the allocated bytes and enqueue as necessary.
+            state.write(chunk, writer);
+        } while (maxBytes > 0 && !queue.isEmpty());
+
+        return totalStreamableBytes > 0;
+    }
+
+    private State state(Http2Stream stream) {
+        return checkNotNull(stream, "stream").getProperty(stateKey);
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class State {
+        final Http2Stream stream;
+        int streamableBytes;
+        boolean enqueued;
+        boolean wasEnqueued;
+
+        private State(Http2Stream stream) {
+            this.stream = stream;
+        }
+
+        void updateStreamableBytes(int newStreamableBytes, boolean hasFrame) {
+            int delta = newStreamableBytes - streamableBytes;
+            if (delta != 0) {
+                streamableBytes = newStreamableBytes;
+                totalStreamableBytes += delta;
+            }
+            if (hasFrame) {
+                // It's not in the queue but has data to send, add it.
+                enqueue();
+            }
+        }
+
+        /**
+         * Write any allocated bytes for the given stream and updates the streamable bytes,
+         * assuming all of the bytes will be written.
+         */
+        void write(int numBytes, Writer writer) throws Http2Exception {
+            if (numBytes == 0 && streamableBytes > 0) {
+                // Don't write a needless empty frame. Just add it back to the queue.
+                enqueue();
+                return;
+            }
+
+            // Update the streamable bytes, assuming that all the bytes will be written.
+            int newStreamableBytes = streamableBytes - numBytes;
+            updateStreamableBytes(newStreamableBytes, newStreamableBytes > 0);
+
+            try {
+                // Write the allocated bytes.
+                writer.write(stream, numBytes);
+            } catch (Throwable t) {
+                throw connectionError(INTERNAL_ERROR, t, "byte distribution write error");
+            }
+        }
+
+        void enqueue() {
```

Comment:
Not sure why ... the class is already private and it's called from the main class.